apps/freeablo/faworld/gamelevel.cpp

```diff
@@ -192,7 +192,7 @@ namespace FAWorld
             Misc::Point targetPoint = point + Misc::Point{xOffset, yOffset};
             if (targetPoint.x >= 0 && targetPoint.x < width() && targetPoint.y >= 0 && targetPoint.y < height())
             {
-                if (isPassable(targetPoint, nullptr))
+                if (isPassable(targetPoint, nullptr) && (additionalConstraints == nullptr || additionalConstraints(targetPoint)))
```

Comment:
Instead of adding this special callback, it would be nice to refactor the class layout a bit.
Ultimately, it would be nice to have a common base class to both Actor and Missile, and a lot of the Actor logic could be moved into this base class (say, GameObject maybe). Then the constraint of no two Actors colliding becomes no two GameObjects colliding. Missile could be a subcless of GameObject, and there could also be a Portal subclass of GameObject (it doesn't really make sense for a portal to be a Missile).
What do you think? Ofc, this is not necessary for this PR, it is alrady big enough as is.

Reply:
That sounds perfect.

Even with the town portal I felt it wasn't quite right. For example if you click a point directly on the other side of a portal you should get routed around it, currently you'll walk straight into it and get teleported. It should be using the "target" stuff already built in for items/actors/doors, but it's currently not that easy to do. Having a common base class (i.e. `GameObject`) with things like a "targetable" trait would make things much easier, and even help tidy up some of the existing codebase.

I agree that a portal definitely isn't a missile. It's easy to get dragged down the same implementation path as the original game when trying to re-use their assets and tables from their exe.

Yep, that'd definitely be it's own PR. I can have a look into it after this, unless you'd like to take it on yourself?

For this PR I can either remove town portal and add it back later, or just make a note that the `additionalConstraints` parameter needs to be removed after the mentioned rework is complete.

Reply:
> Yep, that'd definitely be its own PR. I can have a look into it after this, unless you'd like to take it on yourself?

If you're interested, I'm very happy to let you handle it. I When there's a provisional PR open I might add some early feedback or make some small modifications myself though.

> For this PR I can either remove town portal and add it back later, or just make a note that the additionalConstraints parameter needs to be removed after the mentioned rework is complete?

A comment is fine for now I think. Before version 1.0, having a sub-optimal feature is always better than not having it at all :)